File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
const rightSideView = function(root) {
    if (!root) return [];
    
    const q = [root],
          result = [];
    
    while (q.length) {
        let last = null;
        const qLen = q.length;
        
        for (let i = 0; i < qLen; i++) {
            const node = q.shift();
            last = node.val;
            if (node.left) q.push(node.left);
            if (node.right) q.push(node.right);
        }
        
        result.push(last);
    }
    
    return result;
};

File: medium/199-binary-tree-right-side-view/199-binary-tree-right-side-view.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def right_side_view(root)
    res = []
    return res if !root
    q = Queue.new
    q << root
    
    while !q.empty? do 
        last = nil
        q.size.times do 
            node = q.pop
            last = node.val
            q << node.left if node.left
            q << node.right if node.right
        end
        res << last
    end
    
    res 
end